### DISCUSSION TITLE
Linker Errors when I try to include imagemagick in my project on windows 11?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5633] [Help] [answered]
It's been couple of hours just to make the project compile.

- I downloaded image magick 7 from the official website.
- Copied the `include` and `lib` folders from the installation folder and added them to my c++ project:
![image](https://user-images.githubusercontent.com/49842031/194709579-6ac6acc1-0b20-4b43-906a-3bf282033813.png)
- tried to link these files in my `CMakeLists.txt`:
```
cmake_minimum_required(VERSION 3.23)

project(demo-app)

set(CMAKE_CXX_STANDARD 23)

file(GLOB SOURCE_FILES src/*.cpp)

add_executable(demo ${SOURCE_FILES})

include_directories(src)

#ImageMagick
add_definitions( -DMAGICKCORE_QUANTUM_DEPTH=16 )
add_definitions( -DMAGICKCORE_HDRI_ENABLE=0 )
include_directories(deps/ImageMagick/include)
target_link_libraries(demo 
    "D:/dev_haidar/c++/demo/deps/ImageMagick/lib/CORE_RL_Magick++_.lib"
    "D:/dev_haidar/c++/demo/deps/ImageMagick/lib/CORE_RL_MagickCore_.lib"
    "D:/dev_haidar/c++/demo/deps/ImageMagick/lib/CORE_RL_MagickWand_.lib"
)
```

- When I run the app from vscode, the linker can't find the implementation of the header files:
```
[main] Building folder: demo demo
[build] Starting build
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" --build d:/dev_haidar/c++/demo/build --config Debug --target demo -j 10 --
[build] MSBuild version 17.3.1+2badb37d1 for .NET Framework
[build] main.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __thiscall Magick::Image::Image(void)" (__imp_??0Image@Magick@@QAE@XZ) referenced in function _main [D:\dev_haidar\c++\demo\build\demo.vcxproj]
[build] main.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: virtual __thiscall Magick::Image::~Image(void)" (__imp_??1Image@Magick@@UAE@XZ) referenced in function _main [D:\dev_haidar\c++\demo\build\demo.vcxproj]
[build] D:\dev_haidar\c++\demo\build\Debug\demo.exe : fatal error LNK1120: 2 unresolved externals [D:\dev_haidar\c++\demo\build\demo.vcxproj]
[proc] The command: "C:\Program Files\CMake\bin\cmake.EXE" --build d:/dev_haidar/c++/demo/build --config Debug --target demo -j 10 -- exited with code: 1 and signal: null
[build] Build finished with exit code 1
[main] Failed to prepare executable target with name "undefined"
```

- My app is so simple:
```
#include <iostream>
#include <Magick++.h>

using namespace std;
using namespace Magick;

int main() {
    cout << "Hello, World123!" << endl;
    Image image;
    return 0;
}
```

Please tell me how can I link the `.lib` files so my app can run?

**Comment by snibgo** [comment]
I don't use cmake, so I am guessing.

You try to use libraries "CORE_RL_Magick++_.lib" etc, but with "cmake.EXE --config Debug".

In library names, RL=release, DB=debug. So you are trying to mix debug and release.

**Comment by Haidar0096** [comment, marked as answer]
I solved it, I am still getting an error at runtime, but at least the app runs for now.

 First install [ImageMagick-7.1.0-49-Q16-x64-dll.exe](https://imagemagick.org/archive/binaries/ImageMagick-7.1.0-49-Q16-x64-dll.exe), then I get the `include` and `lib` folders from it, then add them to my project under `Dependencies./ImageMagick`, and this is my `CMakeLists`:

```
cmake_minimum_required(VERSION 3.23)

project(demo-app)

set(CMAKE_CXX_STANDARD 23)

set(CMAKE_EXPORT_COMPILE_COMMANDS ON)

file(GLOB SOURCE_FILES src/*.cpp)

add_executable(demo ${SOURCE_FILES})

include_directories(src)

#ImageMagick
add_definitions( -DMAGICKCORE_QUANTUM_DEPTH=16 )
add_definitions( -DMAGICKCORE_HDRI_ENABLE=0 )
include_directories(deps/ImageMagick/include)
target_link_directories(demo PRIVATE deps/ImageMagick/lib)
target_link_libraries(demo 
    "CORE_RL_Magick++_"
    "CORE_RL_MagickCore_"
    "CORE_RL_MagickWand_"
)
```

Now the following app compiles and runs:

```
#include <Magick++.h>
#include <iostream>

using namespace std;
using namespace Magick;

int main() {
	cout << "Hello, World!" << endl;
	string inputFileName =
		"D:\\magick\\pic2.jpg";
	string outputFileName = "D:\\magick\\pic_resized.png";
	int width = 800;
	int height = 600;

	try{
		Image image;
		image.read(inputFileName);
		Geometry geometry(width, height);
		geometry.aspect(false);
		image.resize(geometry);
		image.write(outputFileName);
		Blob b;
		image.write(&b);
		cout << "data size is " << b.length() << endl;
	}
	catch (...) {
		cout << "error occurred";
	}
	return 0;
}
```

I get an error on the line `image.read(inputFileName);`, but this is another problem. At least the app now compiles and runs.

**Reply by ArhumJain** [reply]
I am having the exact same issue as you were. However, I replicated your answer and it still does not work for me. Any insights?

**Reply by Haidar0096** [reply]
you need to tell me more about thr error you are getting and about your setup so i may be able to help, and btw, I abandoned magick++ and went to use magickwand which is a C api as it is more  suitable in my case  because dart ffi which Iam working with works easier with C apis

**Reply by ArhumJain** [reply]
I copied your code and your `CMakeLists.txt` as well as copying over the lib and include directories from the [ImageMagick-7.1.0-49-Q16-x64-dll.exe](https://imagemagick.org/archive/binaries/ImageMagick-7.1.0-49-Q16-x64-dll.exe) installation. I am using `g++ 8.1.0` and `cmake 3.22-rc2`
Here is my file structure:
![image](https://user-images.githubusercontent.com/75106472/194802620-834cdcb1-0e3d-4989-bdb7-120863223812.png)

My error:
![image](https://user-images.githubusercontent.com/75106472/194802702-976f8550-1b77-46c5-b93e-d69f66cef2d4.png)

**Reply by Haidar0096** [reply]
try using clang++ to compile it. I was using clang++ I remember to compile it.
also are you using cmake extension to start the compilation,? It can help you choose the compiler ( of course you should download it if you don't have it )

**Reply by ArhumJain** [reply]
Yes, I'm using the VS Code cmake extension to compile it. I just tried installing `LLVM` and tried using `clang++` however I get this error:
![image](https://user-images.githubusercontent.com/75106472/194804169-14b1a09d-659f-4d1d-a212-ea9655efba09.png)

`iostream` which should be part of the standard library, is not found.

**Reply by Haidar0096** [reply]
did you select the clang compiler using cmake select kit? I still see mingw in the output.

**Comment by snibgo** [comment]
It may help other people if you say what your build command now is. Have you removed "--config Debug"?

**Reply by Haidar0096** [reply]
no i didn't remove it, I said in the comment above what I changed: the dependency files, and the cmakelists.

As for how I'm running,I am using vscode's extension for cmake (official one):
`ctrl + shift + A` >`Cmake clean rebuild`> `Cmake Debug`
